File: metadata/mill-metadata-service/src/main/java/io/qpointz/mill/metadata/api/DtoMapper.java
package io.qpointz.mill.metadata.api;

import io.qpointz.mill.metadata.api.dto.MetadataEntityDto;
import io.qpointz.mill.metadata.domain.MetadataEntity;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Mapper between domain entities and DTOs.
 */
@Component
public class DtoMapper {
    
    /**
     * Convert domain entity to DTO with merged facets for scope.
     */
    public MetadataEntityDto toDto(MetadataEntity entity, String scope) {
        // Get merged facets for the specified scope
        Map<String, Object> mergedFacets = new HashMap<>();
        for (String facetType : entity.getFacets().keySet()) {
            entity.getFacet(facetType, scope, Object.class)
                .ifPresent(data -> mergedFacets.put(facetType, data));
        }
        
        return MetadataEntityDto.builder()
            .id(entity.getId())
            .type(entity.getType())
            .schemaName(entity.getSchemaName())
            .tableName(entity.getTableName())
            .attributeName(entity.getAttributeName())
            .createdAt(entity.getCreatedAt())
            .updatedAt(entity.getUpdatedAt())
            .createdBy(entity.getCreatedBy())
            .updatedBy(entity.getUpdatedBy())
            .facets(mergedFacets)
            .build();
    }
    
    /**
     * Convert DTO to domain entity.
     */
    public MetadataEntity toEntity(MetadataEntityDto dto) {
        MetadataEntity entity = new MetadataEntity();
        entity.setId(dto.getId());
        entity.setType(dto.getType());
        entity.setSchemaName(dto.getSchemaName());
        entity.setTableName(dto.getTableName());
        entity.setAttributeName(dto.getAttributeName());
        entity.setCreatedAt(dto.getCreatedAt());
        entity.setUpdatedAt(dto.getUpdatedAt());
        entity.setCreatedBy(dto.getCreatedBy());
        entity.setUpdatedBy(dto.getUpdatedBy());
        
        // Convert facets map back to scoped structure
        if (dto.getFacets() != null) {
            Map<String, Map<String, Object>> scopedFacets = new HashMap<>();
            dto.getFacets().forEach((facetType, data) -> {
                Map<String, Object> scopeMap = new HashMap<>();
                scopeMap.put("global", data);
                scopedFacets.put(facetType, scopeMap);
            });
            entity.setFacets(scopedFacets);
        }
        
        return entity;
    }
}

File: metadata/mill-metadata-service/src/main/kotlin/io/qpointz/mill/metadata/api/DtoMapper.kt
package io.qpointz.mill.metadata.api

import io.qpointz.mill.metadata.api.dto.MetadataEntityDto
import io.qpointz.mill.metadata.domain.MetadataEntity
import org.springframework.stereotype.Component

/** Maps domain entities to REST DTOs and back. */
@Component
class DtoMapper {

    /** Converts domain entity to scope-resolved API DTO. */
    fun toDto(entity: MetadataEntity, scope: String): MetadataEntityDto {
        val mergedFacets = mutableMapOf<String, Any?>()
        for (facetType in entity.facets.keys) {
            entity.getFacet(facetType, scope, Any::class.java)
                .ifPresent { mergedFacets[facetType] = it }
        }
        return MetadataEntityDto(
            id = entity.id,
            type = entity.type,
            schemaName = entity.schemaName,
            tableName = entity.tableName,
            attributeName = entity.attributeName,
            createdAt = entity.createdAt,
            updatedAt = entity.updatedAt,
            createdBy = entity.createdBy,
            updatedBy = entity.updatedBy,
            facets = mergedFacets
        )
    }

    /** Converts API DTO into domain entity with default `global` facet scope. */
    fun toEntity(dto: MetadataEntityDto): MetadataEntity {
        val entity = MetadataEntity(
            id = dto.id,
            type = dto.type,
            schemaName = dto.schemaName,
            tableName = dto.tableName,
            attributeName = dto.attributeName,
            createdAt = dto.createdAt,
            updatedAt = dto.updatedAt,
            createdBy = dto.createdBy,
            updatedBy = dto.updatedBy
        )
        dto.facets?.forEach { (facetType, data) ->
            entity.facets.getOrPut(facetType) { mutableMapOf() }["global"] = data
        }
        return entity
    }
}
